cpp/src/arrow/columnar_format.cc

```diff
@@ -0,0 +1,1034 @@
+  impl_->columns_[i].values = values;
+}
+
+int ColumnMap::Find(const std::shared_ptr<Field>& field) const {
+  int i = impl_->Find(field);
+  return (i >= 0) ? i : kFieldNotFound;
+}
+
+int ColumnMap::size() const {
+  return static_cast<int>(impl_->columns_.size());
+}
+
+std::shared_ptr<Field> ColumnMap::field(int i) const {
+  return impl_->columns_[i].field;
+}
+
+void ColumnMap::Get(int i, std::shared_ptr<Int16Array>* rep_levels,
+                           std::shared_ptr<Int16Array>* def_levels,
+                           std::shared_ptr<Array>* values) const {
+  if (rep_levels) {
+    *rep_levels = impl_->columns_[i].rep_levels;
+  }
+  if (def_levels) {
+    *def_levels = impl_->columns_[i].def_levels;
+  }
+  if (values) {
+    *values = impl_->columns_[i].values;
+  }
+}
+
+const std::shared_ptr<Int16Array>& ColumnMap::rep_levels(int i) const {
+  return impl_->columns_[i].rep_levels;
+}
+
+const std::shared_ptr<Int16Array>& ColumnMap::def_levels(int i) const {
+  return impl_->columns_[i].def_levels;
+}
+
+const std::shared_ptr<Array>& ColumnMap::values(int i) const {
+  return impl_->columns_[i].values;
+}
+
+
+// Utility functions and classes
+namespace {
+
+template<typename data_type>
+struct supports_basic_append {
+  static constexpr bool value =
+    has_c_type<data_type>::value ||
+    std::is_same<data_type, BooleanType>::value ||
+    std::is_base_of<FixedSizeBinaryType, data_type>::value;
+};
+
+/// Copy single value from array to builder, generic version
+template<typename data_type>
+typename std::enable_if<supports_basic_append<data_type>::value, Status>::type
+CopyValue(const typename TypeTraits<data_type>::ArrayType& array, int64_t i,
+          typename TypeTraits<data_type>::BuilderType* builder) {
+  return builder->Append(array.Value(i));
+}
+
+
+/// Copy single value from array to builder, specialization for BinaryType
+template<typename data_type>
+typename std::enable_if<std::is_base_of<BinaryType, data_type>::value, Status>::type
+CopyValue(const typename TypeTraits<data_type>::ArrayType& array, int64_t i,
+          typename TypeTraits<data_type>::BuilderType* builder) {
+  int32_t length;
+  const uint8_t* value = array.GetValue(i, &length);
+  return builder->Append(value, length);
+}
+
+
+#define PRIMITIVE_CASE_LIST \
+  CASE(TimestampType) \
+  CASE(BooleanType) \
+  CASE(Int8Type) \
+  CASE(UInt8Type) \
+  CASE(Int16Type) \
+  CASE(UInt16Type) \
+  CASE(Int32Type) \
+  CASE(UInt32Type) \
+  CASE(Int64Type) \
+  CASE(UInt64Type) \
+  CASE(FloatType) \
+  CASE(DoubleType) \
+  CASE(BinaryType) \
+  CASE(StringType) \
+  CASE(FixedSizeBinaryType) \
+  CASE(Date32Type) \
+  CASE(Date64Type) \
+  CASE(Time32Type) \
+  CASE(Time64Type)
+
+} // anonymous namespace
+
+
+// Shredder implementation
+namespace {
+
+class ShredNode {
+ public:
+  ShredNode(const std::shared_ptr<Field>& field, MemoryPool* pool)
+      : field_(field),
+        rep_levels_builder_(std::make_shared<Int16Builder>(pool)),
```

Comment:
Can these ever be null?  I think you are better off using a buffer builder directly for performance reasons.